sched/queue/memory/simple.go

```diff
@@ -3,40 +3,134 @@ package memory
 // In-memory Scheduler Queue
 
 import (
-	"fmt"
 	"github.com/scootdev/scoot/sched"
 	"github.com/scootdev/scoot/sched/queue"
+	"strconv"
 	"time"
 )
 
-func NewSimpleQueue() (queue.Queue, chan queue.WorkItem) {
+func NewSimpleQueue(capacity int) queue.Queue {
 	q := &simpleQueue{}
-	q.outCh = make(chan queue.WorkItem, 1)
-	q.inCh = make(chan enqueueReq)
-	go q.loop()
-	return q, q.outCh
+	q.reqCh = make(chan queueReq)
+	q.outCh = make(chan queue.WorkItem)
+
+	st := &simpleQueueState{reqCh: q.reqCh, outCh: q.outCh, capacity: capacity}
+	go st.loop()
+	return q
 }
 
 type simpleQueue struct {
-	nextID int
-	inCh   chan enqueueReq
-	outCh  chan queue.WorkItem
+	// Immutable state, read by many goroutines
+	reqCh chan queueReq
+	outCh chan queue.WorkItem
+}
+
+type simpleQueueState struct {
+	// Mutable state, only read/written by its goroutine
+	reqCh     chan queueReq
+	outCh     chan queue.WorkItem
+	dequeueCh chan struct{}
+	items     []*simpleWorkItem
+	nextID    int
+	capacity  int
+}
+
+func (s *simpleQueueState) loop() {
+	for !s.done() {
+		s.iter()
+
+	}
+	close(s.outCh)
+}
+
+func (s *simpleQueueState) done() bool {
+	// We are done once our inputs are closed and we have no items left to send
+	return s.reqCh == nil && len(s.items) == 0
 }
 
-type simpleWorkItem sched.Job
+func (s *simpleQueueState) iter() {
```

Comment:
If you have a bunch of requests waiting in L62: 'case req, ok := <-s.reqCh:' due to slow durability or such, then the 'items' array will not change responsively and will not accurately reflect the current load.

In terms of the nil channel, I figured it out but I had to convince myself that the logic was ok and ended up having to look up this particular behavior (well actually I just tried it out on the playground). This is the perfect place to comment that we continually reassign outCh to avoid an 'if' statement (or maybe check if null before assigning?) and that nil channel is ok initially because we have nothing to send on the first iteration and want it to fail/block.